docs/enhancements/add-external-providers.md

```diff
@@ -0,0 +1,182 @@
+providing the user with clear UI to see its provisioned machines joining the cluster, and lowering installation
+resources requirements.
+
+Assisted Installer is also ideal for mixed env configurations.
+** The RHV provider supports installing a mixed env where the Control-Plain (Master) nodes are deployed on top of RHV
+and Compute (Worker) nodes are Bare-Metal nodes.
+
+### Goals
+
+- Create a Provider interface, and hooks in the assisted installer code which separates the assisted installer core code
+ from the provider-specific code.
+- Add clear documentation for each provider specific function.
+- Document steps to install a provider cluster with assisted installer
+
+### Non-Goals
+
+- Support IPI installations.
+- Integrating all relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)
+into the assisted installer code - this will require a large refactor of the assisted installer code.
+- Remove Baremetal from not appropriate stuct/functions names(example in open questions).
+ 
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a RHV/VSphere admin I want to create an Openshift cluster on my RHV engine by using the assisted installer.
+
+#### Story 2
+
+As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer in which
+Control-Plain (Master) nodes are deployed on top of RHV and Compute (Worker) nodes are Bare-Metal nodes.
+
+#### Story 3
+
+As an Openshift developer working on platform X I want to add support for installing Openshift with assisted installer
+on top of platform X.
+
+### Implementation Details/Notes/Constraints
+
+#### Changes to the assisted-service project:
+
+##### Step 1: Detect platform specific logic
+
+Go over the assisted installer flow and detect parts of the code which require provider specific logic.
+
+##### Step 2: Create a Go interface
+
+We should create a Go interface that will contain each provider specific logic in a function
+for each provider to implement.
+
+For example, prototype:
+
+```go
+// Provider contains functions which are required to support installing on a specific platform.
+type Provider interface {
+	// Name returns the name of the platform.
+	Name() string
+	// UpdatePlatformInventoryData sets all platform specific inventory data on the updates map.
+	UpdatePlatformInventoryData(params installer.UpdateClusterParams, cluster *common.Cluster, updates map[string]interface{}) error
+    // SetPlatformInstallConfig sets all platform specific install config parameters
+    SetPlatformInstallcfgValues(platform *platformVsphere, clusterPlatform *models.VspherePlatform)
+    //...
+}
+```
+
+##### Step 3: Implement registry
+
+Implement the registry pattern so providers can register with their names.
+When we will detect the a certain provider then we will register it.
+If no provider is detected then we register the Bare-Metal provider.
```

Comment:
This section is unclear.  Is this describing how the service understands which platforms are available, or how the user chooses which platform to use?  I assume the first.
We will have the bare-metal and none platforms in any case (even if the user is running on virt, in case they don't want/need platform integration).

Reply:
Neither of them actually.
The "registry pattern" is a design pattern that is used a lot in openshift.
Here is an example from the api-server:
- Plugins Register function: https://github.com/openshift/openshift-apiserver/blob/98786f917ffc7d3dc3b05893f405970b87a419b9/vendor/k8s.io/apiserver/pkg/admission/plugins.go#L70-L86
- Usage: https://github.com/openshift/openshift-apiserver/blob/98786f917ffc7d3dc3b05893f405970b87a419b9/pkg/build/apiserver/admission/secretinjector/admission.go#L24-L30

Here is an example that is closer to our usecase, from a PR we abandoned:
- Register definition: https://github.com/openshift/installer/pull/4517/commits/11dc0bcf5555caea832cda52f6c52b64590f6de6#diff-dbd907695a0fe9e8040355b09537c81ac2afd21ad450a2e16468b495ac9d862dR34-R36
- Usage: https://github.com/openshift/installer/pull/4517/commits/21c2ec8c0a1f1450da404083981b5c47d829a3a2#diff-cb477a5c4cf6ed9e8ca2b7c1e817d98737d8bda0879cd30182129b2cc44c6ffeR10

it will be used inside the providerDetector method to register the provider once it is detected

Reply:
I think that's what I meant by "describing how the service understands which platforms are available"?
Each provider registers its name and some method that the service can execute to determine if the platform is available for the current hosts that are discovered.  For example, the RHV plugin with register "RHV" and a method that checks that all of the masters are running on RHV.  At that point, if the method returns True, the service will allow the user to select that platform.  If the user selects it, then the other plugin methods can be executed, such as validations, configuration, etc.

Reply:
Amm yep that is what I meant, one thing that is not clear to me (due to my lack of AI project knowledge probably) how will the user select the oVirt platform if there is no UI changes?
BTW I think we can move this to Open Question 2 since it is related to it